File: app/src/main/java/com/example/teletypesha/netCode/NetServerController.java
package com.example.teletypesha.netCode;

import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.os.Binder;
import android.os.IBinder;
import android.util.Log;

import androidx.annotation.Nullable;

import com.example.teletypesha.activitys.MainActivity;
import com.example.teletypesha.crypt.Crypt;

import java.io.Serializable;
import java.net.URI;
import java.net.URISyntaxException;
import java.sql.Timestamp;
import java.text.SimpleDateFormat;
import java.util.Arrays;
import java.util.Base64;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.CompletableFuture;

import tech.gusavila92.websocketclient.WebSocketClient;

public class NetServerController extends Service implements Serializable {
    private static int k = 0;
    public static String s = "95.165.27.159";

    public static WebSocketClient webSocketClient;
    private static Map<Integer, OnMessageReceived> listeners = new HashMap<>();

    public void AddNewChat(String chatId, String chatPassword) {
    }

    public interface OnMessageReceived {
        void onMessage(String[] parts) throws Exception;
    }

    // Binder, предоставленный клиентам
    private final IBinder binder = new LocalBinder();

    // Класс, используемый для клиентского интерфейса Binder
    public class LocalBinder extends Binder {
        public NetServerController getService() {
            // Возвращаем экземпляр NetServerController,
            // чтобы клиенты могли вызывать публичные методы
            return NetServerController.this;
        }
    }

    public static void setOnMessageReceivedListener(int id, OnMessageReceived listener) {
        listeners.put(id, listener);
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        createWebSocketClient();
        return START_STICKY; // Сервис будет перезапущен, если система его уничтожит
    }

    public void createWebSocketClient() {
        Log.i("WebSocket", "Try createWebSocketClient");
        URI uri;
        try {
            // Connect to local host
            uri = new URI("ws://"+s+":17825/");
        }
        catch (URISyntaxException e) {
            e.printStackTrace();
            return;
        }

        webSocketClient = new WebSocketClient(uri) {
            @Override
            public void onOpen() {
                Log.i("WebSocket", "Session is starting");
                SendUnregistredRequest("Hello World!");
            }

            @Override
            public void onTextReceived(String s) {
                Log.i("WebSocket", "Message received");
                // Обработка полученного сообщения
                String[] parts = s.split(" ");
                try {
                    int id = Integer.parseInt(parts[0]);
                    OnMessageReceived listener = listeners.get(id);
                    if (listener != null) {
                        listener.onMessage(Arrays.copyOfRange(parts, 1, parts.length));
                        listeners.remove(id);
                    }

                }
                catch (Exception e){
                    Log.i("WebSocket", s);
                }
            }

            @Override
            public void onBinaryReceived(byte[] data) {
            }

            @Override
            public void onPingReceived(byte[] data) {
            }

            @Override
            public void onPongReceived(byte[] data) {
            }

            @Override
            public void onException(Exception e) {
                Log.e("WebSocket", Objects.requireNonNull(e.getMessage()));
            }

            @Override
            public void onCloseReceived() {
                Log.i("WebSocket", "Closed ");
            }
        };

        webSocketClient.setConnectTimeout(10000);
        webSocketClient.setReadTimeout(60000);
        webSocketClient.enableAutomaticReconnection(5000);
        webSocketClient.connect();
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return binder;
    }

    private static void SendRequest(int id, String requestWord, String string){
        if(webSocketClient != null)
            webSocketClient.send("/sql " + requestWord + " " + String.valueOf(id) + " " + string);
    }

    private static void SendUnregistredRequest(String string){
        webSocketClient.send(string);
    }

    private static int GetK(){
        k++;
        if(k >= 1000000){
            k = 0;
        }
        return k;
    }

    // -------------------------------------------------------------------------------------------

    // Запросы к серверу

    // -------------------------------------------------------------------------------------------

    // Создание нового чата
    public static CompletableFuture<String> CreateNewChat(String chatPassword, boolean isPrivacy) {
        CompletableFuture<String> future = new CompletableFuture<>();
        int requestId = GetK();

        setOnMessageReceivedListener(requestId , new OnMessageReceived() {
            @Override
            public void onMessage(String[] parts) {
                if (parts.length > 0) {
                    future.complete(parts[0]);
                }
            }
        });

        Log.i("WebSocket", "CreateNewChat");
        SendRequest(requestId, "ChatCreate", isPrivacy + " " + String.valueOf(chatPassword));

        return future;
    }

    public static CompletableFuture<String> AddUserToChat(String publicKey, String idChat, String chatPassword) {
        CompletableFuture<String> future = new CompletableFuture<>();
        int requestId = GetK();

        setOnMessageReceivedListener(requestId , new OnMessageReceived() {
            public void onMessage(String[] parts) {
                if (parts.length > 0) {
                    future.complete(parts[0]);
                } else {
                    future.complete(null);
                }
            }
        });

        Log.i("WebSocket", "addUserToChat");
        SendRequest(requestId, "addUserToChat", publicKey + " " + idChat + " " + chatPassword);

        return future;
    }

    // Отправить сообщение
    public static CompletableFuture<String> SendMessage(byte[] msg, int idSender, Timestamp timeMsg) {
        CompletableFuture<String> future = new CompletableFuture<>();
        int requestId = GetK();

        setOnMessageReceivedListener(requestId , new OnMessageReceived() {
            public void onMessage(String[] parts) {
                if (parts.length > 0) {
                    future.complete(parts[0]);
                } else {
                    future.complete(null);
                }
            }
        });

        Log.i("WebSocket", "SendMessage");
        String timeWithoutMilliseconds = (new SimpleDateFormat("yyyy-MM-dd HH:mm:ss")).format(timeMsg);
        SendRequest(requestId, "SendMessage", String.valueOf(idSender) + " " + String.valueOf(timeWithoutMilliseconds) + " " + Base64.getEncoder().encodeToString(msg));

        return future;
    }

    // Вернуть сообщения, которые больше idMsg
    public static CompletableFuture<String[]> GetMessages(String str) {
        CompletableFuture<String[]> future = new CompletableFuture<>();
        int requestId = GetK();

        setOnMessageReceivedListener(requestId, new OnMessageReceived() {
            @Override
            public void onMessage(String[] parts) {
                if (parts.length > 0) {
                    if (parts.length > 1) {
                        future.complete(parts);
                    } else {
                        future.complete(null);
                    }
                }
            }
        });

        Log.i("WebSocket", "GetMessages");
        SendRequest(requestId, "GetMessages", str);

        return future;
    }

    // Удаление сообщения
    public static CompletableFuture<String> DeleteMessage(int idSender, int idMsg) {
        CompletableFuture<String> future = new CompletableFuture<>();
        int requestId = GetK();

        setOnMessageReceivedListener(requestId, new OnMessageReceived() {
            @Override
            public void onMessage(String[] parts) {
                if (parts.length > 0) {
                    if (parts.length > 1) {
                        future.complete(parts[1]);
                    } else {
                        future.complete(null);
                    }
                }
            }
        });

        Log.i("WebSocket", "DeleteMessage");
        SendRequest(requestId, "DeleteMessages", String.valueOf(idSender) + " " + String.valueOf(idMsg));

        return future;
    }

    // Изменение сообщения
    public static CompletableFuture<String> RefactorMessage(int idMsg, int idSender, byte[] msg) {
        CompletableFuture<String> future = new CompletableFuture<>();
        int requestId = GetK();
        setOnMessageReceivedListener(requestId, new OnMessageReceived() {
            @Override
            public void onMessage(String[] parts) {
                if (parts.length > 0) {
                    future.complete(parts[0]);
                } else {
                    future.complete(null);
                }
            }
        });

        Log.i("WebSocket", "RefactorMessage");
        SendRequest(requestId, "RefactorMessage", String.valueOf(idMsg) + " " + String.valueOf(idSender) + " " + Base64.getEncoder().encodeToString(msg));

        return future;
    }

    public static CompletableFuture<String> Login(String log, String pass, Context context) throws Exception {
        CompletableFuture<String> future = new CompletableFuture<>();
        int requestId = GetK();
        setOnMessageReceivedListener(requestId, new OnMessageReceived() {
            @Override
            public void onMessage(String[] parts) {
                if (parts.length > 0) {
                    future.complete(parts[0]);
                } else {
                    future.complete(null);
                }
            }
        });

        Log.i("WebSocket", "Login");
        SendRequest(requestId, "GetUserData", Crypt.CriptUser(log, pass, context, false));

        return future;
    }

    public static CompletableFuture<Boolean> Register(String log, String pass, Context context) throws Exception {
        CompletableFuture<Boolean> future = new CompletableFuture<>();
        int requestId = GetK();
        setOnMessageReceivedListener(requestId, new OnMessageReceived() {
            @Override
            public void onMessage(String[] parts) {
                if (parts.length > 0) {
                    future.complete(Boolean.valueOf(parts[0]));
                } else {
                    future.complete(false);
                }
            }
        });

        Log.i("WebSocket", "Register");
        String str = Crypt.CriptUser(log, pass, context, true);
        SendRequest(requestId, "AddUserData", str);

        return future;
    }
}

File: Flutter/tele_vibe/lib/GettedData/netServerController.dart
import 'dart:async';
import 'dart:convert';

import 'package:flutter/material.dart';
import 'package:web_socket_channel/status.dart' as status;
import 'package:web_socket_channel/web_socket_channel.dart';

class NetServerController with WidgetsBindingObserver {
  static final String s = "95.165.27.159";
  static WebSocketChannel? webSocketChannel;
  static int k = 0;
  static Map<int, Function(List<String>)> listeners = {};

  static final NetServerController _instance = NetServerController._internal();

  factory NetServerController() {
    return _instance;
  }

  NetServerController._internal();

  void start() {
    WidgetsBinding.instance.addObserver(this);
    if (webSocketChannel == null || webSocketChannel!.closeCode != null) {
      createWebSocketClient();
    }
  }

  void createWebSocketClient() {
    final uri = Uri.parse("ws://$s:17825/");
    webSocketChannel = WebSocketChannel.connect(uri);

    webSocketChannel!.stream.listen((message) {
      onTextReceived(message);
    }, onDone: () {
      print('WebSocket closed');
      start(); // Restart connection if closed
    }, onError: (error) {
      print('WebSocket error: $error');
      start(); // Restart connection on error
    });

    sendUnregistredRequest("Hello World!");
  }

  void onTextReceived(String message) {
    List<String> parts = message.split(" ");
    try {
      int id = int.parse(parts[0]);
      var listener = listeners[id];
      if (listener != null) {
        listener(parts.sublist(1));
        listeners.remove(id);
      }
    } catch (e) {
      print('Error: $e');
    }
  }

  void setOnMessageReceivedListener(int id, Function(List<String>) listener) {
    listeners[id] = listener;
  }

  void sendRequest(int id, String requestWord, String message) {
    if (webSocketChannel != null) {
      webSocketChannel!.sink.add("/sql $requestWord $id $message");
    }
  }

  void sendUnregistredRequest(String message) {
    if (webSocketChannel != null) {
      webSocketChannel!.sink.add(message);
    }
  }

  int getK() {
    k++;
    if (k >= 1000000) {
      k = 0;
    }
    return k;
  }

  // Requests to the server

  Future<String> createNewChat(String chatPassword, bool isPrivacy) async {
    Completer<String> completer = Completer<String>();
    int requestId = getK();

    setOnMessageReceivedListener(requestId, (parts) {
      if (parts.isNotEmpty) {
        completer.complete(parts[0]);
      }
    });

    sendRequest(requestId, "ChatCreate", "$isPrivacy $chatPassword");
    return completer.future;
  }

  Future<String> addUserToChat(String publicKey, String idChat, String chatPassword) async {
    Completer<String> completer = Completer<String>();
    int requestId = getK();

    setOnMessageReceivedListener(requestId, (parts) {
      if (parts.isNotEmpty) {
        completer.complete(parts[0]);
      } else {
        completer.complete(null);
      }
    });

    sendRequest(requestId, "addUserToChat", "$publicKey $idChat $chatPassword");
    return completer.future;
  }

  Future<String> sendMessage(List<int> msg, int idSender, DateTime timeMsg) async {
    Completer<String> completer = Completer<String>();
    int requestId = getK();
    String timeWithoutMilliseconds = timeMsg.toIso8601String().split('.').first;

    setOnMessageReceivedListener(requestId, (parts) {
      if (parts.isNotEmpty) {
        completer.complete(parts[0]);
      } else {
        completer.complete(null);
      }
    });

    sendRequest(requestId, "SendMessage", "$idSender $timeWithoutMilliseconds ${base64Encode(msg)}");
    return completer.future;
  }

  Future<List<String>> getMessages(String str) async {
    Completer<List<String>> completer = Completer<List<String>>();
    int requestId = getK();

    setOnMessageReceivedListener(requestId, (parts) {
      if (parts.isNotEmpty) {
        completer.complete(parts);
      } else {
        completer.complete([]);
      }
    });

    sendRequest(requestId, "GetMessages", str);
    return completer.future;
  }

  Future<String> deleteMessage(int idSender, int idMsg) async {
    Completer<String> completer = Completer<String>();
    int requestId = getK();

    setOnMessageReceivedListener(requestId, (parts) {
      if (parts.length > 1) {
        completer.complete(parts[1]);
      } else {
        completer.complete(null);
      }
    });

    sendRequest(requestId, "DeleteMessages", "$idSender $idMsg");
    return completer.future;
  }

  Future<String> refactorMessage(int idMsg, int idSender, List<int> msg) async {
    Completer<String> completer = Completer<String>();
    int requestId = getK();

    setOnMessageReceivedListener(requestId, (parts) {
      if (parts.isNotEmpty) {
        completer.complete(parts[0]);
      } else {
        completer.complete(null);
      }
    });

    sendRequest(requestId, "RefactorMessage", "$idMsg $idSender ${base64Encode(msg)}");
    return completer.future;
  }

  Future<String> login(String log, String pass) async {
    Completer<String> completer = Completer<String>();
    int requestId = getK();

    setOnMessageReceivedListener(requestId, (parts) {
      if (parts.isNotEmpty) {
        completer.complete(parts[0]);
      } else {
        completer.complete(null);
      }
    });

    sendRequest(requestId, "GetUserData", log + " " + pass); // Assuming Crypt.CriptUser is replaced with log + " " + pass for simplicity
    return completer.future;
  }

  Future<bool> register(String log, String pass) async {
    Completer<bool> completer = Completer<bool>();
    int requestId = getK();

    setOnMessageReceivedListener(requestId, (parts) {
      if (parts.isNotEmpty) {
        completer.complete(parts[0] == 'true');
      } else {
        completer.complete(false);
      }
    });

    sendRequest(requestId, "AddUserData", log + " " + pass); // Assuming Crypt.CriptUser is replaced with log + " " + pass for simplicity
    return completer.future;
  }

  @override
  void didChangeAppLifecycleState(AppLifecycleState state) {
    if (state == AppLifecycleState.resumed) {
      // Restart WebSocket connection if needed when the app resumes
      start();
    }
  }

  void dispose() {
    WidgetsBinding.instance.removeObserver(this);
    webSocketChannel?.sink.close(status.goingAway);
  }
}

void main() {
  WidgetsFlutterBinding.ensureInitialized();
  NetServerController().start();

  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter WebSocket Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  void initState() {
    super.initState();
    NetServerController().start();
  }

  @override
  void dispose() {
    NetServerController().dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Flutter WebSocket Demo'),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text('WebSocket Connection Example'),
          ],
        ),
      ),
    );
  }
}